add check validity of edm::Handle in HLTDQM modules

mtosi opened the pull request:
as pointed out in https://github.com/cms-sw/cmssw/pull/24155#issuecomment-413131664 
and spotted / seen also in https://hypernews.cern.ch/HyperNews/CMS/get/tier0-Ops/1978/1/1.html

some modules make the job crashing (!!!)
I've added the check of the handle validity

mtosi committed:
add check validity of Handles

fabiocos commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

fabiocos commented:
@mtosi as we have discussed directly:

1) the fix prevents the crash, verified with Configuration/dataProcessing

python RunExpressProcessing.py --scenario ppEra_Run2_2018 --fevt --dqmio --global-tag 102X_dataRun2_Express_v2 --lfn /store/t0streamer/Data/Express/000/321/879/run321879_ls0315_streamExpress_StorageManager.dat

using as a source

process.source = cms.Source("NewEventStreamFileReader",
    fileNames = cms.untracked.vstring('/store/t0streamer/Data/Express/000/321/879/run321879_ls0315_streamExpress_StorageManager.dat')
)

2) I suggest to add LogWarnings for a limited number of checks so as to avoid a silent failures but also to avoid filling log files with failure messages;

3) please forward-port to master;

4) again this is a work-around, not the ultimate solution (a proper management of the DQM express sequences discussed since long time).

cmsbuild commented:
+1
Tested at: 3410be5e0718eacff44afd88f5ef291affc87739
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24445/30222/summary.html

cmsbuild commented:
Comparison job queued.

mtosi committed:
add LogWarning

Update BPHMonitor.cc

Update BPHMonitor.cc

Update HTMonitor.cc

Update METMonitor.cc

cmsbuild commented:
Pull request #24445 was updated. @kmaeshima, @cmsbuild, @andrius-k, @jfernan2, @schneiml can you please check and sign again.

mtosi commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

schneiml commented:
@fabiocos as you mentioned, this does not really fix the problem, but a proper fix will be much harder/riskier.

IIRC @mtosi said that they don't need any HLT DQM in StreamExpress anyways, so we could just disable that entirely. Except we can't currently, since StreamExpress is using the default relval sequence...

Another thing that would be interesting is if we can reproduce this crash using `runTheMatrix`. Fixing WF 1001.2 to use the "proper", default sequence as used in real Express did not do it, but it is based on 2017 data. I wonder what else changed that causes this problem to appear in T0 express.

cmsbuild commented:
+1
Tested at: 552f4a4a8d8c2c1cfc0a248d27b2025ccd7fcb92
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24445/30228/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24445/30228/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 31
 - DQMHistoTests: Total histograms compared: 2985378
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2985186
 - DQMHistoTests: Total skipped: 190
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 30 files compared)
 - Checked 129 log files, 14 edm output root files, 31 DQM output files

fabiocos commented on DQMOffline/Trigger/plugins/BPHMonitor.cc:
```diff
@@ -345,17 +355,37 @@ void BPHMonitor::analyze(edm::Event const& iEvent, edm::EventSetup const& iSetup
   edm::Handle<reco::BeamSpot> beamSpot;
   iEvent.getByToken( bsToken_,  beamSpot);
+  if ( !beamSpot.isValid() ) {
+    if (warningPrinted4token_[0]) {
```
@mtosi in order to work, this check should have an inverted logic, i.e. if (!warningPrinted4token_[0]) { since the flag is initialized to false (verified with DataProcessing).

cmsbuild commented:
Pull request #24445 was updated. @kmaeshima, @cmsbuild, @andrius-k, @jfernan2, @schneiml can you please check and sign again.

cmsbuild commented:
Pull request #24445 was updated. @kmaeshima, @cmsbuild, @andrius-k, @jfernan2, @schneiml can you please check and sign again.

cmsbuild commented:
Pull request #24445 was updated. @kmaeshima, @cmsbuild, @andrius-k, @jfernan2, @schneiml can you please check and sign again.

cmsbuild commented:
Pull request #24445 was updated. @kmaeshima, @cmsbuild, @andrius-k, @jfernan2, @schneiml can you please check and sign again.

fabiocos commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

fabiocos commented:
@schneiml I believe did a mistake in accepting the original PR, as no further action was taken to fix it until we saw the problem in T0 replay. We may setup relval tests that emulate express 2018, but a Configuration/DataProcessing run of a couple of minutes is enough to see the problem, as was originally discussed. Now this test is passed, and only one instance of the warning message is written.

At this point I want to quickly prevent 10_2_X to fail. We are discussing since quite sometime of a proper adjustment of the express DQM sequences but this does not seem a short term goal. 

In this specific case I assume that a task should be added to the sequence so as to produce in unscheduled mode the ak8PFPuppiSoftDrop collections that are needed, and that are not produced so far (the module is not present in the standard jet reco sequences as far as I can see)
@perrotta @slava77 could you please confirm my understanding?

schneiml commented:
> In this specific case I assume that a task should be added to the sequence so as to produce in unscheduled mode the ak8PFPuppiSoftDrop collections that are needed, and that are not produced so far (the module is not present in the standard jet reco sequences as far as I can see)
@perrotta @slava77 could you please confirm my understanding?

As I understand, it would make more sense to not run DQM on things that are not produced in the given job. As mentioned earlier (but not yet confirmed by @mtosi ) HLT does not even care about the additional plots in StreamExpress.

fabiocos commented:
@schneiml I agree, but the definition of the express sequences is a ball in your field :-)

cmsbuild commented:
+1
Tested at: 50e410b993cda6ddfd6072b0decbaa33225043d3
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24445/30239/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24445/30239/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 31
 - DQMHistoTests: Total histograms compared: 2985378
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2985185
 - DQMHistoTests: Total skipped: 190
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 30 files compared)
 - Checked 129 log files, 14 edm output root files, 31 DQM output files

mtosi commented:
as reported @ the DQM meeting [1] and stated many times
TSG does not need anything on the Express PD

<img width="905" alt="image" src="https://user-images.githubusercontent.com/4946419/45029707-dbb56700-b049-11e8-9d6d-885946d0c8d5.png">


[1]
https://indico.cern.ch/event/744063/contributions/3073967/attachments/1687931/2715077/180717_STEAM.pdf

mtosi committed:
moved softdrop modules to WithRECO sequence

fix logic

cmsbuild commented:
Pull request #24445 was updated. @kmaeshima, @cmsbuild, @andrius-k, @jfernan2, @schneiml can you please check and sign again.

fabiocos commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 542d552c05b2eae683a31805b2b22ed4754460bb
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24445/30263/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/b8c8b43b5c0d3c9e9e33340d9acfcc0272e74c5a
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24445/30263/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24445/30263/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24445/30263/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 31
 - DQMHistoTests: Total histograms compared: 2985378
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2985187
 - DQMHistoTests: Total skipped: 190
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 6.536 KiB( 30 files compared)
 - DQMHistoSizes: changed ( 150.0,... ): 3.268 KiB        HLT/B2G
 - Checked 129 log files, 14 edm output root files, 31 DQM output files

cmsbuild commented:
Pull request #24445 was updated. @kmaeshima, @cmsbuild, @andrius-k, @jfernan2, @schneiml can you please check and sign again.

cmsbuild commented:
Pull request #24445 was updated. @kmaeshima, @cmsbuild, @andrius-k, @jfernan2, @schneiml can you please check and sign again.

mtosi commented:
I've implemented the logic discussed with Fabio:
- check the InputTag is not empty
- only if it is not, then skip the event (not fill the plot)

and then I rebase for limiting / squeezing the number of commits

mtosi commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

mtosi commented:
@fabiocos could this PR be integrated and a patch release w/ this be built ?
TSG asked to PdmV to produce samples for the 102X validation w/ data, and we need this PR in

please, let me know if there is anything to do
and thanks !

cmsbuild commented:
+1
Tested at: fede58c3fb9fe9218e9b6ac3123110c88d6044ab
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24445/30282/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24445/30282/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 31
 - DQMHistoTests: Total histograms compared: 2985378
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2985187
 - DQMHistoTests: Total skipped: 190
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 6.536 KiB( 30 files compared)
 - DQMHistoSizes: changed ( 150.0,... ): 3.268 KiB        HLT/B2G
 - Checked 129 log files, 14 edm output root files, 31 DQM output files

fabiocos commented:
@jfernan2 @schneiml @andrius-k could you please check and sign it in case?

andrius-k commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_10_2_X IBs (tests are also fine) and once validation in the development release cycle CMSSW_10_3_X is complete. This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

fabiocos commented:
+1
